File: apps/api/app/schemas/audit.py
"""Pydantic schemas for audit results."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class Severity(str, Enum):
    INFO = "info"
    WARNING = "warning"
    HIGH = "high"
    CRITICAL = "critical"


class DimensionScores(BaseModel):
    completeness: float = Field(ge=0, le=100)
    uniqueness: float = Field(ge=0, le=100)
    validity: float = Field(ge=0, le=100)
    consistency: float = Field(ge=0, le=100)
    documentation: float = Field(ge=0, le=100)


class QualityIssue(BaseModel):
    id: str
    audit_run_id: str
    column_name: str | None = None
    issue_type: str
    dimension: str
    severity: Severity
    message: str
    affected_rows_count: int = 0
    affected_rows_sample: list[dict[str, Any]] = Field(default_factory=list)
    recommendation: str
    created_at: datetime


class ColumnProfile(BaseModel):
    id: str
    audit_run_id: str
    column_name: str
    inferred_type: str
    null_count: int
    null_rate: float
    distinct_count: int
    distinct_rate: float
    sample_values: list[Any] = Field(default_factory=list)
    min_value: Any | None = None
    max_value: Any | None = None
    most_common_values: list[dict[str, Any]] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class DataDictionaryField(BaseModel):
    id: str
    audit_run_id: str
    column_name: str
    inferred_type: str
    nullable: bool
    description_suggestion: str
    example_values: list[Any] = Field(default_factory=list)
    quality_notes: str = ""


class AuditRun(BaseModel):
    id: str
    dataset_name: str
    uploaded_filename: str
    row_count: int
    column_count: int
    overall_score: float
    completeness_score: float
    uniqueness_score: float
    validity_score: float
    consistency_score: float
    documentation_score: float
    created_at: datetime
    status: str = "completed"
    executive_recommendation: str = ""
    columns: list[ColumnProfile] = Field(default_factory=list)
    issues: list[QualityIssue] = Field(default_factory=list)
    data_dictionary: list[DataDictionaryField] = Field(default_factory=list)
    report_markdown: str = ""
    datapackage: dict[str, Any] = Field(default_factory=dict)


class AuditSummary(BaseModel):
    id: str
    dataset_name: str
    uploaded_filename: str
    row_count: int
    column_count: int
    overall_score: float
    dimensions: DimensionScores
    top_issues: list[QualityIssue]
    executive_recommendation: str
    created_at: datetime
    status: str

File: apps/web/src/types/audit.ts
export type Severity = "info" | "warning" | "high" | "critical";

export type DimensionScores = {
  completeness: number;
  uniqueness: number;
  validity: number;
  consistency: number;
  documentation: number;
};

export type QualityIssue = {
  id: string;
  audit_run_id: string;
  column_name: string | null;
  issue_type: string;
  dimension: string;
  severity: Severity;
  message: string;
  affected_rows_count: number;
  affected_rows_sample: Record<string, unknown>[];
  recommendation: string;
  created_at: string;
};

export type ColumnProfile = {
  id: string;
  audit_run_id: string;
  column_name: string;
  inferred_type: string;
  null_count: number;
  null_rate: number;
  distinct_count: number;
  distinct_rate: number;
  sample_values: unknown[];
  min_value: unknown;
  max_value: unknown;
  most_common_values: { value: string; count: number }[];
  warnings: string[];
};

export type DataDictionaryField = {
  id: string;
  audit_run_id: string;
  column_name: string;
  inferred_type: string;
  nullable: boolean;
  description_suggestion: string;
  example_values: unknown[];
  quality_notes: string;
};

export type AuditRun = {
  id: string;
  dataset_name: string;
  uploaded_filename: string;
  row_count: number;
  column_count: number;
  overall_score: number;
  completeness_score: number;
  uniqueness_score: number;
  validity_score: number;
  consistency_score: number;
  documentation_score: number;
  created_at: string;
  status: string;
  executive_recommendation: string;
  columns: ColumnProfile[];
  issues: QualityIssue[];
  data_dictionary: DataDictionaryField[];
  report_markdown: string;
  datapackage: Record<string, unknown>;
};

export type AuditSummary = {
  id: string;
  dataset_name: string;
  uploaded_filename: string;
  row_count: number;
  column_count: number;
  overall_score: number;
  dimensions: DimensionScores;
  top_issues: QualityIssue[];
  executive_recommendation: string;
  created_at: string;
  status: string;
};

export type DemoDataset = {
  id: string;
  title: string;
  description: string;
  filename: string;
  available: boolean;
  kind?: "synthetic" | "public_sample";
};
